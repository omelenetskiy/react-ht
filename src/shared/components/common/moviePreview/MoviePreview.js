import React, { lazy, Suspense, Component, Fragment } from "react";
import { Helmet } from "react-helmet";
import { connect } from "react-redux";
import { withRouter } from "react-router";
import {
  actionMovieFetch,
  actionMovieFetchIfNeeded
} from "../../../store/actions/movie";
import { getMovieState, getMoviesState } from "../../../store/selectors";

import MovieList from "../movieList/MovieList";
const MovieCard = lazy(() => import("./movieCard/MovieCard"));

class MoviePreview extends Component {
  static fetchData({ store, params }) {
    store.dispatch(actionMovieFetch({ id: parseInt(params.id) }));
  }

  fetchIfNeeded() {
    this.props.dispatch(
      actionMovieFetchIfNeeded({ id: parseInt(this.props.match.params.id) })
    );
  }

  componentDidMount() {
    this.fetchIfNeeded();
  }

  componentDidUpdate(prevProps) {
    if (prevProps.match.params.id !== this.props.match.params.id) {
      this.fetchIfNeeded();
    }
  }

  render() {
    const { moviesData } = this.props;
    const { movies } = moviesData;
    const { movie } = this.props.movieData;
    const date = Date.parse(movie.release_date);
    const parsedDate = new Date(date).getFullYear();
    return (
      <Fragment>
        <Helmet>
          <title>{`${movie.title}`}</title>
        </Helmet>

        <Suspense fallback={<div>Loading...</div>}>
          <MovieCard movie={movie} parsedDate={parsedDate} />
        </Suspense>

        <section className="home">
          <MovieList movies={movies} />
        </section>
      </Fragment>
    );
  }
}

const mapStateToProps = state => {
  return {
    movieData: getMovieState(state),
    moviesData: getMoviesState(state)
  };
};

export default withRouter(connect(mapStateToProps)(MoviePreview));
