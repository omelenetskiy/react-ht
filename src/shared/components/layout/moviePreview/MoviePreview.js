import React, { Component, Fragment } from "react";
import { Helmet } from "react-helmet";
import { connect } from "react-redux";
import {
  actionMovieFetch,
  actionMovieFetchIfNeeded
} from "../../../store/actions/movie";

class Movie extends Component {
  static fetchData({ store, params }) {
    return store.dispatch(actionMovieFetch({ id: parseInt(params.id) }));
  }

  fetchIfNeeded() {
    this.props.dispatch(
      actionMovieFetchIfNeeded({ id: parseInt(this.props.match.params.id) })
    );
  }

  componentDidMount() {
    this.fetchIfNeeded();
  }

  componentDidUpdate(prevProps) {
    if (prevProps.match.params.id !== this.props.match.params.id) {
      this.fetchIfNeeded();
    }
  }

  render() {
    const { movie } = this.props.movieData;
    const date = Date.parse(movie.release_date);
    const parsedDate = new Date(date).getFullYear();
    return (
      <Fragment>
        <Helmet>
          <title>{`${movie.title}`}</title>
        </Helmet>
        {this.props.movieData.loading ? (
          "Loading..."
        ) : (
          <article className="movie">
            <div className="card">
              <img src={movie.poster_path} className="card-img" alt="..." />
              <div className="card-body">
                <h1 className="card-title">
                  {`${movie.title} `}
                  <span className="badge badge-pill badge-danger">
                    {movie.vote_average}
                  </span>
                </h1>
                <p className="card-text">
                  <small>{movie.tagline}</small>
                </p>
                <p className="card-text">
                  <small>{`${parsedDate}`}</small>{" "}
                  <small>{`${movie.runtime} min`}</small>
                </p>
                <p className="card-text">{movie.overview}</p>
              </div>
            </div>
          </article>
        )}
      </Fragment>
    );
  }
}

function moviesState(state) {
  return {
    movieData: state.movie
  };
}

export default connect(moviesState)(Movie);
