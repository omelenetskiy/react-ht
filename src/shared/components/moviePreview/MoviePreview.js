import React, { Component, Fragment } from 'react';
import { Helmet } from 'react-helmet';
import { connect } from 'react-redux';
import { withRouter } from 'react-router';
import { actionMovieFetch } from '../../store/actions/movie';
import { getMovieState, getMoviesState } from '../../store/selectors';

import MovieList from '../movieList/MovieList';
import MovieCard from './movieCard/MovieCard';

export class MoviePreview extends Component {
  static fetchData({ store, params }) {
    store.dispatch(actionMovieFetch({ id: parseInt(params.id, 10) }));
  }

  componentDidMount() {
    this.fetchIfNeeded();
  }

  componentDidUpdate(prevProps) {
    const { match } = this.props;
    if (prevProps.match.params.id !== match.params.id) {
      this.fetchIfNeeded();
    }
  }

  fetchIfNeeded() {
    const { dispatch, match } = this.props;
    dispatch(actionMovieFetch({ id: parseInt(match.params.id, 10) }));
  }

  render() {
    const { moviesData, movieData } = this.props;
    const { movies } = moviesData;
    const { movie } = movieData;
    const date = Date.parse(movie.release_date);
    const parsedDate = new Date(date).getFullYear();
    return (
      <Fragment>
        <Helmet>
          <title>{`${movie.title}`}</title>
        </Helmet>

        <MovieCard movie={movie} parsedDate={parsedDate} />

        <section className="home">
          <MovieList movies={movies} />
        </section>
      </Fragment>
    );
  }
}

const mapStateToProps = (state) => ({
  movieData: getMovieState(state),
  moviesData: getMoviesState(state),
});

export default withRouter(connect(mapStateToProps)(MoviePreview));
