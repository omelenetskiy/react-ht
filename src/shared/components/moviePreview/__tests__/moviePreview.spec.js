import React from 'react';
import configureStore from 'redux-mock-store';
import { shallow } from 'enzyme';
import { Provider } from 'react-redux';
import toJson from 'enzyme-to-json';
import 'jest-enzyme';
import 'jest-styled-components';

import MoviePreview from '../MoviePreview';

const initialState = {};
const mockStore = configureStore();
let store;

const props = {
  movie: {
    title: 'title',
    release_date: 'release_date',
    poster_path: 'poster_path',
    vote_average: 'vote_average',
    tagline: 'tagline',
    runtime: 'runtime',
    overview: 'overview',
  },
  parsedDate: 'parsedDate',
};

describe('MovieCard', () => {
  beforeEach(() => {
    store = mockStore(initialState);
  });
  it('should render correctly', () => {
    const wrapper = shallow(
      <Provider store={store}>
        <MoviePreview {...props} />
      </Provider>
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render Loading...', () => {
    const wrapper = shallow(
      <Provider store={store}>
        <MoviePreview />
      </Provider>
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
