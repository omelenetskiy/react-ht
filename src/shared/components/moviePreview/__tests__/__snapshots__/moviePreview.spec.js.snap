// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieCard should render Loading... 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "storeState": Object {},
    }
  }
>
  <Connect(MoviePreview) />
</ContextProvider>
`;

exports[`MovieCard should render correctly 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "storeState": Object {},
    }
  }
>
  <Connect(MoviePreview)
    movie={
      Object {
        "overview": "overview",
        "poster_path": "poster_path",
        "release_date": "release_date",
        "runtime": "runtime",
        "tagline": "tagline",
        "title": "title",
        "vote_average": "vote_average",
      }
    }
    parsedDate="parsedDate"
  />
</ContextProvider>
`;
